src/geodesy.jl

```diff
@@ -0,0 +1,447 @@
+            end
+        end
+        @inbounds output[i] = max(current_min,mask[i])
+        for Δi in lower_deltaoffest # examine neighborhoods
+            ii = i + Δi
+            if checkbounds(Bool, R, ii) #check that we are in the image
+                @inbounds if output[ii]>output[i] && output[ii]>mask[ii]
+                    enqueue!(queue, i)
+                end
+            end
+        end
+    end
+    # Loop until all pixel have been examined 
+    while !isempty(queue)
+        curr_idx = dequeue!(queue)
+        for Δi in deltaoffsets # examine neighborhoods
+            ii = curr_idx + Δi
+            if checkbounds(Bool, R, ii) #check that we are in the image
+                @inbounds if output[ii] > output[curr_idx] && mask[ii] != output[ii] 
+                    output[ii] = max(output[curr_idx],mask[ii])
+                    enqueue!(queue, ii)
+                end
+            end
+        end
+    end
+    return output
+end
+
+#specialization for binary case
+function overbuild(marker::AbstractArray{<:Union{Bool,AbstractGray{Bool}},N}, mask::AbstractArray{<:Union{Bool,AbstractGray{Bool}},N}, connectivity::AbstractArray{Bool})  where {N}
+    marker_inv =  map(!, marker)
+    mask_inv = map(!, mask)
+    output_inv= underbuild(marker_inv,mask_inv,connectivity)
+    return map(!, output_inv)
+end
+
+"""
+    regional_maxima(image, connectivity=[]) -> Array
+
+Determines all *regional maxima* of the `image`.
+
+# Details
+The *regional maximum* is defined as the connected set of pixels that have the
+same value, which is greater than the values of all pixels in direct
+neighborhood of the set.
+This implementation is faster than maxtree approach if maxtree not precomputed
+# Arguments
+- `image::AbstractArray{T, N}`: where {T<:ImageCore.NumberLike , N} the ``N``-dimensional input image
+- `connectivity::AbstractArray{Bool}`: the neighborhood connectivity.
+"""
+regional_maxima(image::AbstractArray{T, N}, connectivity::AbstractArray{Bool}) where {T<:ImageCore.NumberLike , N} = extract_regional_extrema(image, connectivity, isgreater, isequal ) 
+
+"""
+    regional_minima(image, connectivity=[]) -> Array
+
+Determines all *regional minima* of the `image`.
+
+# Details
+The *regional minima* is defined as the connected set of pixels that have the
+same value, which is lower than the values of all pixels in direct
+neighborhood of the set.
+
+This implementation is faster than maxtree approach if maxtree not precomputed
+
+# Arguments
+- `image::AbstractArray{T, N}`: where {T<:ImageCore.NumberLike , N} the ``N``-dimensional input image
+- `connectivity::AbstractArray{Bool}`: the neighborhood connectivity.
+"""
+regional_minima(image::AbstractArray{T, N}, connectivity::AbstractArray{Bool}) where {T<:ImageCore.NumberLike , N} = extract_regional_extrema(image, connectivity, isless, isequal ) 
+
+
+function neighborhood(A::AbstractArray{T,N}, connectivity::AbstractArray{Bool}) where {T,N}
```

Comment:
Since this isn't performance bottleneck, how about the shorter version:

```julia
neighborhood(connectivity) = append!(upper_neighborhood(connectivity), lower_neighborhood(connectivity))

upper_neighborhood(x) = _neighborhood(i->x[i], CartesianIndices(x))
lower_neighborhood(x) = _neighborhood(i->x[i], reverse(CartesianIndices(x)))

function _neighborhood(f, R)
    for i in R
        i == center && break
        if f(i)
            push!(...)
        end
    end
    return out
end
```

Reply:
+1